File: leetcode2/0109. Convert Sorted List to Binary Search Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        if not head:
            return None
        if not head.next:
            return TreeNode(head.val)
        
        node=ListNode(0)
        node.next=head
        pslow,pfast=node,node
        while pfast and pfast.next and pfast.next.next:
            pslow=pslow.next
            pfast=pfast.next.next
        pright=pslow.next
        pslow.next=None
        pleft=head
        
        root=TreeNode(pright.val)
        root.left=self.sortedListToBST(pleft)
        root.right=self.sortedListToBST(pright.next)
        
        return root
        

File: archives/C#/0109. Convert Sorted List to Binary Search Tree.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode SortedListToBST(ListNode head) {
        if(head==null){
            return null;
        }
        if(head.next==null){
            return new TreeNode(head.val);
        }
        ListNode node=new ListNode(0);
        node.next=head;
        ListNode pslow=node;
        ListNode pfast=node;
        while(pfast.next!=null && pfast.next.next!=null){
            pslow=pslow.next;
            pfast=pfast.next.next;
        }
        TreeNode treeNode=new TreeNode(pslow.next.val);
        treeNode.right=SortedListToBST(pslow.next.next);
        pslow.next=null;
        treeNode.left=SortedListToBST(head);
        return treeNode;
        
    }
}
